### DISCUSSION TITLE
Using babashka with PHP: guestbook example

### DISCUSSION THREAD

**Discussion by borkdude** [borkdude/blog#21] [Posts]
Using babashka with PHP: guestbook example

**Comment by rlhk** [comment]
The example is very helpful for my recent review & migration of a PHP web app, which was built by an inexperienced team. The web app feature set is quite sizeable. Surely it comes with the classical PHP mess of a mix of OOP & imperative styles. Using bb + mysql pod allows a very simple bridging path towards a Clojure solution. A lot of PHP code simply becomes redundant. 

Technically, I use `exec`. One fun discovery is that when $long_str is too long, `exec("echo $long_str >> /tmp/dev.log")` won't work. In my case, just a few hundred chars.

I wish a pod of HoneySQL exists. Maybe I could try make one later.

**Reply by borkdude** [reply]
@rlhk HoneySQL doesn't need a pod since it works from source with babashka :)

Just put it in your bb.edn and it should work.

**Reply by rlhk** [reply]
Problem is that my remote server doesn't have java installed. 
When I have `(deps/add-deps '{:deps {honeysql/honeysql {:mvn/version "1.0.444"}}})`
It complains:
```
----- Error --------------------------------------------------------------------
Type:     java.lang.Exception
Message:  Couldn't find 'java'. Please set JAVA_HOME.
...
```

**Reply by borkdude** [reply]
You can git clone honeysql and then add it to the classpath manually.

**Reply by rlhk** [reply]
@borkdude I tried git clone honeysql but unable get it work.
Without honeysql, I could call my script from within php like:
```
$cmd = "QID=service-id QARGS='[$service_id]'".
                 " bb --classpath ".DIR_ROOT."clojure/bb/ --main fm.db-query";
```
Instead of `bb.edn`, guest `--classpath` is easier as the call requires absolute path. Not sure though.

Now my folder structure with honeysql cloned is:
```
<php-project-root>
  clojure
    bb
      fm
        db_query.clj
      honeysql
        src/honey
        ...
      honey <-- Symbolic link to honeysql/src/honey
```
I always get `Could not find namespace: honeysql.core.` error no matter how I move the honeysql folder around. 
Tried to have multiple class path option like: `--classpath ./a/b/c:./x/y/z` and no luck.

So few questions:
1. What's the proper folder structure of git cloned libs? A sample project should be easiest to understand.
2. If `bb.edn` is used. How to specify it's location from `bb ...`? The reason is that `bb ...` might not be involved from the same folder.
3. Libs like honeysql has it's own dependencies, I'm curious to learn how those external libs get pulled by bb if honeysql is successfully loaded.

Thanks !

**Reply by borkdude** [reply]
1 You can specify multiple entries on the classpath separated with `:` (or `;` on Windows). So the classpath would become:
 `".DIR_ROOT."clojure/bb:".DIR_ROOT."clojure/bb/honeysql/src`
2 You can specify the bb.edn with the `--config` argument
3 HoneySQL doesn't have any other deps than clojure, and bb already has clojure built-in.

**Reply by borkdude** [reply]
Ad 3: if HoneySQL would have more deps, then you would just also clone those and add them manually to the classpath as well.

**Reply by rlhk** [reply]
> 1 You can specify multiple entries on the classpath separated with `:` (or `;` on Windows). So the classpath would become:
> `".DIR_ROOT."clojure/bb:".DIR_ROOT."clojure/bb/honeysql/src`
> 2 You can specify the bb.edn with the `--config` argument
> 3 HoneySQL doesn't have any other deps than clojure, and bb already has clojure built-in.

Retried 1 with both relative & abs path, no luck.
For 2. Used a simple bb.edn as `{:paths ["." "./honeysql/src"]}` and invoked bb in the same path. Also tried `--config ...` same could not find namespace: honeysql.core error.

Thanks for the hints anyway.

**Reply by borkdude** [reply]
@rlhk Try putting `(prn (map str (map fs/absolutize (babashka.classpath/split-classpath (babashka.classpath/get-classpath)))))` in your main function and see what it returns. That might show you that the relative paths are not resolved correctly.

**Reply by borkdude** [reply]
Btw, you could also just copy honeysql's sources into your own src directory and then the classpath issue should be solved as well.

**Reply by rlhk** [reply]
> @rlhk Try putting `(prn (map str (map fs/absolutize (babashka.classpath/split-classpath (babashka.classpath/get-classpath)))))` in your main function and see what it returns. That might show you that the relative paths are not resolved correctly.

When `bb.edn` is `{:paths ["." "./honeysql/src"]}`
The printout is  `("<abs-php-project-root>/clojure/./bb/." "<abs-php-project-root>/clojure/./bb/./honeysql/src")`
Seems strange but it actually works without honeysql as the -main fn runs.

Changing `bb.edn` to `{:paths ["" "honeysql/src"]}`
The printout is `("< abs-php-project-root>/clojure/./bb/." "< abs-php-project-root>/clojure/./bb/./honeysql/src")`
Looks different but works the same way as above.

Both give same honeysql.core error when `(require '[honeysql.core :as hsql])` is added to code.

**Reply by rlhk** [reply]
Well. Actually all kinds of setup including the `--classpath` option works. Turns out honeysql has changed the main ns from `honeysql.core` to `honey.sql`, which I copied from https://github.com/babashka/babashka-sql-pods for testing. LOL. 

Guess it's easier for you to modify the example instead of a PR?

**Reply by borkdude** [reply]
@rlhk The require for honeysql is no longer `honeysql.core` but `honey.sql`. See the README of that project.

**Reply by borkdude** [reply]
PR welcome :)

**Reply by rlhk** [reply]
> PR welcome :)

Sure, will do :D
